### DISCUSSION TITLE
Add Temperature Control entity

### DISCUSSION THREAD

**Discussion by hookedonunix** [home-assistant/architecture#923] [Entity Models] [answered]
## Proposal
Add a Temperature Control (or Number Control) entity to Home Assistant to allow devices a common method to implement controllable attributes.

Climate and water heater are currently the only platforms that support setting temperature, but this restricts any device that wishes to get and set temperature to an existing platform implementation. It also makes it impossible for these devices (other than climate) to be controlled with third party services like Google Home and Apple HomeKit. Providing a common entity for this functionality would allow for devices made of smaller and less bespoke entities.

If there is interest in this I will try and flesh out the interface of this better, but following Google I think a good start is:

`NumberControlEntity`

**Attributes**
- `current_value`
- `max_value`
- `min_value`
- `target_value`
- `target_value_high`
- `target_value_low`
- `unit_of_measurement`
- `precision`
- `device_class`
  - `water_heater`
  - `oven`
  - `bathtub`
  - `humidifier`
- `supported_features`
  - `TARGET_VALUE`
  - `TARGET_VALUE_RANGE`

**Services**
- `set_value`
- `turn_on`/`turn_off`

## Background
Found while [adding the Water Heater to the Google Assistant integration](https://github.com/home-assistant/core/pull/93648) where we found the current water heater entity doesn't match the Google device type.

As this was created in my attempt to integrate with Google Assistant, my proposal is primarily based of how Google has chosen to tackle this problem with their [`TemperatureControl`](https://developers.home.google.com/cloud-to-cloud/traits/temperaturecontrol) trait.

Also mentioned [here](https://github.com/home-assistant/architecture/discussions/518#discussioncomment-4189621).

**Comment by frenck** [comment]
You are describing an `number` entity in your attributes and services. Why not use that? It even has a `temperature` device class for it specifically.

../Frenck

**Reply by elupus** [reply]
I could see a reason for more advanced number that support a upper/lower and current value. Ie split this out from climate and water heater.

Would be useful for temperature targets,  humidity targets and other controlled things.

Ie a number target entity, thay can both reflect a setpoint and a current value, that doest need to match setpoint.

**Reply by frenck** [reply]
Right, but in that case the `climate` entity can just be used?

**Reply by elupus** [reply]
Not for a water heater, or hygrostat. Yes we do have entity types for those. But i think there could be a benefit of splitting those rather large entity types. Another use case could be a bang bang controlled water level controller (dont think it exists :)). But any bang bang/pi controlled process could benefit from this type of thing.

**Reply by hookedonunix** [reply]
The climate entity uses another Google Assistant trait called [TemperatureSetting](https://developers.home.google.com/cloud-to-cloud/traits/temperaturesetting), which functions slightly differently and restricts the maximum temperature. Devices such as Ovens, Water Heaters, Vaporizers, and Kettles can operate at higher temperatures and aren't as well suited to climate (at least the way Google has differentiated them).

I'm looking for a way to be able to send them as the correct device type via Google Assistant while getting current temperature and getting/setting target temperature

**Reply by hookedonunix** [reply]
Another thing to add to this entity would be supported features like `TARGET_TEMPERATURE` and `TARGET_TEMPERATURE_RANGE`, which are not currently on Number

**Comment by jeeftor** [comment]
What else besides water heaters would need this?

**Reply by hookedonunix** [reply]
If made to accommodate generic number control entities it could be used as a base for humidifier or hygrostat entities, as suggested above.
Theoretically it could replace the logic present for current and target temps in the climate component, but that would be an impractically large change for only a little benefit.

It would also allow a number of device types that are supported by Google Assistant to be used with Home Assistant in a consistent manner ([bathtub](https://developers.home.google.com/cloud-to-cloud/guides/bathtub), [coffee maker](https://developers.home.google.com/cloud-to-cloud/guides/coffeemaker), [humidifier](https://developers.home.google.com/cloud-to-cloud/guides/humidifier), [oven](https://developers.home.google.com/cloud-to-cloud/guides/oven) [water heater](https://developers.home.google.com/cloud-to-cloud/guides/waterheater), etc.). Currently each integration must add support manually.

For my purposes, I only require a small change to the water heater entity, but as I've seen this feature requested a few times I thought I'd check to see if a more generic approach would be helpful to anyone.

**Comment by frenck** [comment, marked as answer]
We have discussed this again in an architectural meeting/discussion, and decided against it for now.

This is also driven by the fact that we are looking into supporting the concept of actual device and device types in the very near feature, which will change or landscape quite a bit. After that has been shipped, we could maybe reconsider a variant of this one.

For now, closing.

../Frenck
